pkg/volume/azure_dd/attacher.go

```diff
@@ -24,191 +24,168 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/wait"
-	"k8s.io/kubernetes/pkg/cloudprovider"
+	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/util/exec"
-	"k8s.io/kubernetes/pkg/util/keymutex"
 	"k8s.io/kubernetes/pkg/util/mount"
 	"k8s.io/kubernetes/pkg/volume"
-	"k8s.io/kubernetes/pkg/volume/util"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
 )
 
+type azureDiskDetacher struct {
+	plugin *azureDataDiskPlugin
+}
+
 type azureDiskAttacher struct {
-	host          volume.VolumeHost
-	azureProvider azureCloudProvider
+	plugin *azureDataDiskPlugin
 }
 
 var _ volume.Attacher = &azureDiskAttacher{}
+var _ volume.Detacher = &azureDiskDetacher{}
 
-var _ volume.AttachableVolumePlugin = &azureDataDiskPlugin{}
+func (a *azureDiskAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	var lun int
+	var err error
+	volumeSource, err := getVolumeSource(spec)
+	if err != nil {
+		return "", err
+	}
 
-const (
-	checkSleepDuration = time.Second
-)
+	diskName := volumeSource.DiskName
+	hashedDiskUri := a.plugin.diskController.MakeCRC32(strings.ToLower(volumeSource.DataDiskURI))
+	glog.V(4).Infof("azureDisk - attempting to check if disk %s attached to node %s", diskName, nodeName)
 
-// acquire lock to get an lun number
-var getLunMutex = keymutex.NewKeyMutex()
+	isManagedDisk := (*volumeSource.Kind == v1.AzureManagedDisk)
+	attached, lun, err := a.plugin.diskController.IsDiskAttached(hashedDiskUri, string(nodeName), isManagedDisk)
 
-// NewAttacher initializes an Attacher
-func (plugin *azureDataDiskPlugin) NewAttacher() (volume.Attacher, error) {
-	azure, err := getAzureCloudProvider(plugin.host.GetCloudProvider())
 	if err != nil {
-		glog.V(4).Infof("failed to get azure provider")
-		return nil, err
+		glog.Infof("azureDisk - error checking if Azure Disk  (%s) is already attached to  node (%s). Will continue and try attach anyway. err:%v", diskName, nodeName, err)
 	}
 
-	return &azureDiskAttacher{
-		host:          plugin.host,
-		azureProvider: azure,
-	}, nil
-}
+	if err == nil && attached {
+		glog.Warningf("azureDisk - disk already attach: (%s) is already attached to node (%s)", diskName, nodeName)
+		return strconv.Itoa(lun), nil
+	} else {
+		glog.V(4).Infof("azureDisk - Disk(%s) is not  attached to node (%s), will attach", diskName, nodeName)
+	}
 
-// Attach attaches a volume.Spec to an Azure VM referenced by NodeName, returning the disk's LUN
-func (attacher *azureDiskAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
-	volumeSource, err := getVolumeSource(spec)
-	if err != nil {
-		glog.Warningf("failed to get azure disk spec")
-		return "", err
+	err = nil // reset just in case the check for attached earlier failed
+
+	cachingMode := string(*volumeSource.CachingMode)
+	managed := (*volumeSource.Kind == v1.AzureManagedDisk)
+	if managed {
+		lun, err = a.plugin.diskController.AttachManagedDisk(string(nodeName), volumeSource.DataDiskURI, cachingMode)
```

Comment:
nodeName != InstanceID
refer to https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/azure_dd/attacher.go#L76